mtags/src/main/scala-3/scala/meta/internal/pc/Signatures.scala

```diff
@@ -0,0 +1,269 @@
+package scala.meta.internal.pc
+
+import java.{util => ju}
+
+import collection.mutable.ListBuffer
+import dotty.tools.dotc.core.Symbols._
+import dotty.tools.dotc.core.Types._
+import dotty.tools.dotc.core.Contexts.Context
+import dotty.tools.dotc.core.Names.Name
+import dotty.tools.dotc.core.NameKinds.EvidenceParamName
+
+import dotty.tools.dotc.core.Flags._
+
+trait Signatures {
+
+  class ShortenedNames(context: Context) {
+    val history = collection.mutable.Map.empty[Name, ShortName]
+
+    def lookupSymbol(short: ShortName): Type = {
+        import scala.meta.internal.mtags.MtagsEnrichments._
+        context.findRef(short.name)
+    }
+
+    def tryShortenName(short: ShortName)(using Context): Boolean = {
+      history.get(short.name) match {
+        case Some(ShortName(_, other)) => true
+        case None =>
+          val foundTpe = lookupSymbol(short)
+          val syms = List(foundTpe.termSymbol, foundTpe.typeSymbol)
+          val isOk = syms.filter(_ != NoSymbol) match {
+            case Nil => false
+            case founds => founds.exists(_ == short.symbol)
+          }
+          if (isOk) {
+            history(short.name) = short
+            true
+          } else {
+            false
+          }
+      }
+    }
+
+    def tryShortenName(name: Option[ShortName])(using Context): Boolean =
+      name match {
+        case Some(short) => tryShortenName(short)
+        case None => false
+      }
+  }
+
+  /** Compute detail string for CompletionItem#detail, which is also used by
+    * method's signature string in label.
+    * 
+    * for
+    * - method: method signature
+    *   - e.g. `[A: Ordering](x: List[Int]): A`
+    * - class or module: package name that class or module is belonging
+    *   - e.g. ` java.lang`
+    * - otherwise: its type
+    *   - e.g. ` java.lang.String` ` Symbols.Symbol`
+    */
+  def infoString(sym: Symbol, info: Type, history: ShortenedNames)(using Context): String = {
+    sym match {
+      case m if m.is(Method) =>
+        SignaturePrinter(sym, history, info).defaultMethodSignature()
+      case _ if sym.isClass || sym.is(Module) =>
+        val printer = SymbolPrinter()
+        s" ${printer.fullNameString(sym.owner)}"
+      case _ =>
+        val short = shortType(info, history)
+        val printer = SymbolPrinter()
+        s" ${printer.typeString(short)}"
+    }
+  }
+
+  class SignaturePrinter(
+      gsym: Symbol,
+      shortenedNames: ShortenedNames,
+      gtpe: Type,
+      // includeDocs: Boolean,
+      // includeDefaultParam: Boolean = true,
+      // printLongType: Boolean = true
+  )(using Context) {
+    // In case rawParamss is no set, fallback to paramSymss
+    private val paramss = if (gsym.rawParamss.length != 0) gsym.rawParamss else gsym.paramSymss
```

Comment:
sometimes, `rawParamss` not yet set, and fallback to paramSymss
https://github.com/tanishiking/dotty/blob/64aa7cfb0a0a245261c490e268052880a07f5c00/compiler/src/dotty/tools/dotc/core/SymDenotations.scala#L303